pkg/auditlog/publisher.go

```diff
@@ -0,0 +1,158 @@
+package alog
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+// Publisher represents an audit log events publisher that posts events to
+// WorkOS asynchronously.
+type Publisher struct {
+	// The WorkOS api key. It can be found in
+	// https://dashboard.workos.com/api-keys.
+	APIKey string
+
+	// The http.Client that is used to post audit log events to WorkOS. Defaults
+	// to http.Client.
+	Client *http.Client
+
+	// The endpoint used to request Workos. Defaults to
+	// https://api.workos.com/events.
+	Endpoint string
+
+	// The function used to encode in JSON. Defaults to json.Marshal.
+	JSONEncode func(v interface{}) ([]byte, error)
```

Comment:
Do we know when a custom JSON encoder would be used?

Reply:
Yep for performance reason. Go default json encoder is slower and allocates more than other 3rd party serializer. Allowing this to be set can open to a serious optimization for the user.

For example, back on my days in Segment, we implemented a JSON encoder that literally blow away the go standard json encoder in term of performance and resource footprint.

Reply:
Should workos-go provide a better encoder by default if that is the case?

Reply:
I think going with the standard library is a solid choice. We should not enforce the use of a 3rd party library.